src/aspire/source/image.py

```diff
@@ -802,6 +805,221 @@ def save_images(
                 im = self.images[i_start:i_end]
                 im.save(mrcs_filepath, overwrite=overwrite)
 
+    def estimate_signal_mean_energy(
+        self, sample_n=None, support_radius=None, batch_size=512
+    ):
+        """
+        Estimate the signal mean of `sample_n` projections.
+
+        :param sample_n: Number of images used for estimate.
+            Defaults to all images in source.
+        :param support_radius: Pixel radius used for masking signal support.
+            Default of None will compute inscribed circle, `self.L // 2`.
+        :param batch_size: Images per batch, defaults 512.
+        :returns: Estimated signal mean
+        """
+
+        if sample_n is None:
+            sample_n = self.n
+
+        if sample_n > self.n:
+            logger.warning(
+                f"`estimate_signal_mean_energy` sample_n > Source.n: {sample_n} > {self.n}."
+                f" Accuracy may be impaired, setting sample_n=self.n={self.n}"
+            )
+            sample_n = self.n
+
+        mask = support_mask(self.L, support_radius, dtype=self.dtype)
+
+        # mean is estimated batch-wise, compare with numpy
+        # Note, for simulation we are implicitly assuming taking `sample_n` is random,
```

Comment:
To support `from_snr` on the `NoiseAdder` side instead of a sim method, I think we should consider dropping the ability to prescribe `sample_n` subsets. It will introduce more weird logic and more coupling.  Then we can remove the randomization, since it only applies for subsets of images.  

Like I _can_ admit `sample_n` to `NoiseAdder.from_snr`, but then I have to pass it back to the Sim side anyway... yuk.

If a user wants to finely control these things they'll just need to do what the original `Simiulation.from_snr` I wrote did, on their own. Make a sim, measure it however they'd like, compute an snr, and assign to the sim.  Basic functions to measure are all built out now, and if something more exotic is desired, they can start from the basic functions.

Reply:
Ok.